### PR TITLE
Lint and dependencies update

### PR DESCRIPTION
**Pull request by sbalay**
# Summary
- Update dependencies
- Fix linter warnings in base project and in generated React Native project
- Remove usage of deprecated `React.PropTypes` in favour of `prop-types`
- Add fix for conflicts between eslint plugins and config extends. Check [this](https://github.com/airbnb/javascript/tree/master/packages/eslint-config-airbnb#eslint-config-airbnb-1)

### CODE REVIEW

File: generators/app/tasks/installDependencies.js [hunk 1]
```diff
@@ -31,9 +32,37 @@ const DEV_DEPENDENCIES = [
...
+    loadingMessage: `Getting eslint plugins' versions`,
+    successMessage: `Successfuly download plugins version info`,
+    failureMessage: `Error getting info of eslint plugins. Turn verbose mode on for detailed logging`,
+    context: options
+  }).then(({ result }) => {
+    const dependencies = JSON.parse(result);
+    // keep latest if the dependency has different versions. e.g: eslint: '^3.19.0 || ^4.3.0'
+    Object.keys(dependencies).forEach(eachDependency => {
+      const versions = dependencies[eachDependency].split('||');
+      dependencies[eachDependency] = versions.slice(-1)[0].trim();
```
Comment (mvbattan) [inline generators/app/tasks/installDependencies.js]: `dependencies[eachDependency].split('||')` will always be sorted by version, from oldest to newest ?

File: generators/app/tasks/installDependencies.js [hunk 2]
```diff
@@ -53,9 +82,15 @@ module.exports = function installDependencies() {
...
-    });
+  return getLinterPluginVersions(this.projectName, this.options).then(plugins => {
+    const pluginNames = Object.keys(plugins);
+    const fixedDevDeps = DEV_DEPENDENCIES.map(
+      dependency => (pluginNames.includes(dependency) ? `${dependency}@${plugins[dependency]}` : dependency)
+    );
+    return yarnInstall(this.projectName, DEPENDENCIES, this.options)
+      .then(() => yarnInstall(this.projectName, fixedDevDeps, this.options, true))
+      .catch(() => {
+        process.exit(1);
```
Comment (mvbattan) [inline generators/app/tasks/installDependencies.js]: I'd keep this `1` as is, but if we are going to add more exit codes later, it is desirable to extract this into a constants file.

File: generators/app/tasks/runCommand.js [hunk 3]
```diff
@@ -40,6 +43,7 @@ module.exports = function runCommand(options) {
       if (options.timeout) {
         refreshKillTimeout();
       }
+      result += data ? `\n${data.toString()}` : '';
```
Comment (mvbattan) [inline generators/app/tasks/runCommand.js]: I'd use a `join` over an array of strings, but this is acceptable.

File: generators/app/templates/src/redux/drawerHandlers.js [hunk 4]
```diff
@@ -21,9 +21,8 @@ export const actionCreators = {
 export function reducer(state = Immutable({ present: false }), action) {
   switch (action.type) {
     case actions.DRAWER_TOGGLED: {
```
Comment (mvbattan) [inline generators/app/templates/src/redux/drawerHandlers.js]: Add a `namespace` for these actions

File: generators/app/tasks/installDependencies.js [hunk 5]
```diff
@@ -31,9 +32,37 @@ const DEV_DEPENDENCIES = [
...
+      'npm',
+      ['info', 'eslint-config-airbnb@latest', 'peerDependencies', '--json'],
+      { cwd: `${process.cwd()}/${projectName}` }
+    ],
+    loadingMessage: `Getting eslint plugins' versions`,
+    successMessage: `Successfuly download plugins version info`,
+    failureMessage: `Error getting info of eslint plugins. Turn verbose mode on for detailed logging`,
+    context: options
+  }).then(({ result }) => {
+    const dependencies = JSON.parse(result);
```
Comment (mvbattan) [inline generators/app/tasks/installDependencies.js]: Extract this block into a function

File: generators/app/tasks/installDependencies.js [hunk 6]
```diff
@@ -53,9 +82,15 @@ module.exports = function installDependencies() {
   if (this.features.drawerios || this.features.drawerandroid) {
     DEPENDENCIES.push('react-native-drawer');
   }
-  return yarnInstall(this.projectName, DEPENDENCIES, this.options)
-    .then(() => yarnInstall(this.projectName, DEV_DEPENDENCIES, this.options, true))
-    .catch(() => {
-      process.exit(1);
-    });
+  return getLinterPluginVersions(this.projectName, this.options).then(plugins => {
+    const pluginNames = Object.keys(plugins);
```
Comment (mvbattan) [inline generators/app/tasks/installDependencies.js]: Extract this block into a function

File: generators/app/tasks/nextSteps.js [hunk 7]
```diff
@@ -14,8 +14,8 @@ module.exports = function nextSteps() {
 
   // fabric
   console.log(
-    `● Fabric and crashlytics are two ${'MANDATORY'.bold.underline} tools in every react-native project in Wolox.`
-      .cyan
+    `● Fabric and crashlytics are two ${'MANDATORY'.bold
+      .underline} tools in every react-native project in Wolox.`.cyan
   );
   console.log('  Follow these guides in order to complete their installation:'.cyan);
   console.log(`    ${'https://fabric.io/kits/ios/crashlytics/manual-install'.cyan.underline}`);
```
Comment (mvbattan) [inline generators/app/tasks/nextSteps.js]: I'd extract links to a constants file

File: generators/app/tasks/nextSteps.js [hunk 8]
```diff
@@ -30,8 +30,8 @@ module.exports = function nextSteps() {
         .cyan
     );
     console.log(
-      `  Further instructions can be found here: ${'https://facebook.github.io/react-native/docs/pushnotificationios.html'.underline}`
-        .cyan
+      `  Further instructions can be found here: ${'https://facebook.github.io/react-native/docs/pushnotificationios.html'
```
Comment (mvbattan) [inline generators/app/tasks/nextSteps.js]: Same here

File: generators/app/templates/src/redux/authHandlers.js [hunk 9]
```diff
@@ -106,10 +106,10 @@ export function reducer(state = Immutable(defaultState), action) {
 
 export function propTypes() {
```
Comment (mvbattan) [inline generators/app/templates/src/redux/authHandlers.js]: Additional comments for this file:
Put in `actionCreators`only these being exported.
I'd move `PropTypes` to `reducers`
Extract route names
Add namespaces
Put this `actions.js` into `redux/auth`

File: generators/app/templates/src/redux/authHandlers.js [hunk 10]
```diff
@@ -1,6 +1,6 @@
-import React from 'react';
-import Immutable from 'seamless-immutable';
 import { NavigationActions } from 'react-navigation';
+import Immutable from 'seamless-immutable';
+import PropTypes from 'prop-types';
 
 import { setCurrentUser, removeCurrentUser } from '../services/AuthService';
```
Comment (mvbattan) [inline generators/app/templates/src/redux/authHandlers.js]: I'd export `AuthService` as default

File: generators/app/templates/src/redux/drawerHandlers.js [hunk 11]
```diff
@@ -1,5 +1,5 @@
 import Immutable from 'seamless-immutable';
```
Comment (mvbattan) [inline generators/app/templates/src/redux/drawerHandlers.js]: `redux/drawer/actions.js`

File: generators/app/templates/src/redux/drawerHandlers.js [hunk 12]
```diff
@@ -35,5 +34,5 @@ export function reducer(state = Immutable({ present: false }), action) {
 /* ------------- Drawer propTypes ------------- */
 
 export const propTypes = {
```
Comment (mvbattan) [inline generators/app/templates/src/redux/drawerHandlers.js]: Props in reducers ?

**Comment by sbalay** [comment]
@mvbattan I'm working on creating a separate pull request with more changes to make this bootstrap match our latests conventions. Fixes like splitting actions and reducers, and changing the folder structure will be in this new pull request once this gets merged.

File: generators/app/tasks/installDependencies.js (hunk 2 above)
Comment (sbalay) [inline generators/app/tasks/installDependencies.js]: it is unlikely that we will have different error codes

File: generators/app/tasks/runCommand.js (hunk 3 above)
Comment (sbalay) [inline generators/app/tasks/runCommand.js]: Agreed!

File: generators/app/templates/src/redux/drawerHandlers.js (hunk 11 above)
Comment (sbalay) [inline generators/app/templates/src/redux/drawerHandlers.js]: out of the scope of this pull request. will be changed in a new one this week

File: generators/app/templates/src/redux/drawerHandlers.js (hunk 4 above)
Comment (sbalay) [inline generators/app/templates/src/redux/drawerHandlers.js]: out of the scope of this pull request. will be changed in a new one this week

File: generators/app/templates/src/redux/drawerHandlers.js (hunk 12 above)
Comment (sbalay) [inline generators/app/templates/src/redux/drawerHandlers.js]: Holy sheet. Where should we put props declarations...

File: generators/app/templates/src/redux/authHandlers.js (hunk 10 above)
Comment (sbalay) [inline generators/app/templates/src/redux/authHandlers.js]: out of the scope of this pull request. will be changed in a new one this week

File: generators/app/templates/src/redux/authHandlers.js (hunk 9 above)
Comment (sbalay) [inline generators/app/templates/src/redux/authHandlers.js]: out of the scope of this pull request. will be changed in a new one this week

File: generators/app/tasks/installDependencies.js (hunk 5 above)
Comment (sbalay) [inline generators/app/tasks/installDependencies.js]: I don't really agree here. This are only few lines of code that solve a single problem that is in the scope of the function. I'll add some comments to make this clear.

File: generators/app/tasks/installDependencies.js (hunk 1 above)
Comment (sbalay) [inline generators/app/tasks/installDependencies.js]: Fixed!

File: generators/app/templates/src/redux/drawerHandlers.js (hunk 12 above)
Comment (sbalay) [inline generators/app/templates/src/redux/drawerHandlers.js]: out of the scope of the pr btw

**Comment by sbalay** [comment]
@mvbattan ready!

File: generators/app/tasks/installDependencies.js [hunk 13]
```diff
@@ -31,9 +35,38 @@ const DEV_DEPENDENCIES = [
...
+    loadingMessage: `Getting eslint plugins' versions`,
+    successMessage: `Successfuly download plugins version info`,
+    failureMessage: `Error getting info of eslint plugins. Turn verbose mode on for detailed logging`,
+    context: options
+  }).then(({ result }) => {
+    // keep latest if the dependency has different versions. e.g: eslint: '^3.19.0 || ^4.3.0'
+    const dependencies = JSON.parse(result);
+    Object.keys(dependencies).forEach(eachDependency => {
+      const latest = latestSemver(dependencies[eachDependency].match(semverRegex()));
+      dependencies[eachDependency] = `^${latest}`;
```
Comment (mvbattan) [inline generators/app/tasks/installDependencies.js]: 🤔  Does `^next` work ? I'd let it as is anyway 😇

Comment (mvbattan) [inline generators/app/tasks/installDependencies.js]: I'd also rename `latest` to `latestVersion`

**Pull request merged**